File: leetcode/problems/reverse_vowels_of_a_str.rb
# Write a function that takes a string as input and
# reverse only the vowels of a string.
#
# Example 1:
# Given s = "hello", return "holle".
#
# Example 2:
# Given s = "leetcode", return "leotcede".
#
# Note:
# The vowels does not include the letter "y".

# param {String} s
# return {String}

def reverse_vowels(s)
  l, r = 0, s.length - 1

  while l < r
    while l < r && !"aeiou".include?(s[l].downcase)
      l += 1
    end

    while l < r && !"aeiou".include?(s[r].downcase)
      r -= 1
    end

    s[l], s[r] = s[r], s[l]
    l += 1
    r -= 1
  end

  s
end

File: leetcode/javascript_problems/reverse_vowels_of_a_str.js
// Write a function that takes a string as input and
// reverse only the vowels of a string.
//
// Example 1:
// Given s = "hello", return "holle".
//
// Example 2:
// Given s = "leetcode", return "leotcede".
//
// Note:
// The vowels does not include the letter "y".

/**
 * @param {string} s
 * @return {string}
 */

var reverseVowels = (s) => {
  let vowels = "aeiouAEIOU".split('');
  let wordArr = s.split('');
  let vowelsInWord = [];

  for (let i = 0; i < s.length; i++) {
    if (vowels.includes(s[i])) {
      vowelsInWord.push(s[i]);
    }
  }

  for (let i = 0; i < wordArr.length; i++) {
    if (vowels.includes(wordArr[i])) {
      wordArr[i] = vowelsInWord.pop();
    }
  }

  return vowelsInWord.join('');
};
